core/src/main/scala/org/apache/spark/sql/execution/columnar/impl/JDBCSourceAsColumnarStore.scala

```diff
@@ -539,17 +538,27 @@ class JDBCSourceAsColumnarStore(private var _connProperties: ConnectionPropertie
         val region = Misc.getRegionForTable(tableName, true).asInstanceOf[LocalRegion]
         region match {
           case pr: PartitionedRegion =>
-            if (partitionId == -1) {
-              val primaryBucketIds = pr.getDataStore.
-                  getAllLocalPrimaryBucketIdArray
-              // TODO: do load-balancing among partitions instead
-              // of random selection
-              val numPrimaries = primaryBucketIds.size()
-              // if no local primary bucket, then select some other
-              if (numPrimaries > 0) {
-                primaryBucketIds.getQuick(rand.nextInt(numPrimaries))
+            if (partitionId == -1) pr.synchronized {
+              val primaryBuckets = pr.getDataStore.getAllLocalPrimaryBucketRegions
+              // if no local primary bucket, then select a random bucket
+              if (primaryBuckets.isEmpty) {
+                Random.nextInt(pr.getTotalNumberOfBuckets)
               } else {
-                rand.nextInt(pr.getTotalNumberOfBuckets)
+                // select the bucket with smallest size at this point
```

Comment:
size here means number of cached batches. I mean the bucket that is selected for insertion has least number of cached batches not the least size. Mentioning a line in comments/ PR would help.

Reply:
Ok will do. Normally we expect cached batches to be similar in size (except when there is a lot of mix of JDBC single inserts with bulk inserts because rollover uses number of rows for policy). Avoided iteration but perhaps getTotalBytes might be better.